.github: add Copilot instructions for AI-generated code

emdroid opened the pull request:
Add comprehensive coding guidelines for GitHub Copilot to improve quality and consistency of AI-generated code. Instructions cover C++ and Python development with language-specific best practices, build system usage, and testing workflows.

Following GitHub Copilot's standard layout with general instructions in .github/copilot-instructions.md and language-specific files in .github/instructions/ directory using *.instructions.md naming.

No backport: This change is only for developers in master, so it doesn't need to be backported.

avikivity commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+
+# 1. Coding Style and Formatting
+
+* **Auto-Formatting:** All C++ code must be formatted using the `.clang-format` file located in the repository root. This is a non-negotiable rule.
```
Wait what? Will it reformat existing code?

Another non-negotiable rule is to keep patches minimal.

emdroid replied:
Right, it should only eventually reformat the coda that is being changed. Instructed Copilot to update this wording, and to add the requirement for minimal patches.

avikivity commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+
+#### 3\. Modern C++ Usage
+
+* **C++ Standard:** The codebase uses C++20 or newer. Prefer features from these standards, especially coroutines, as they improve code clarity and safety.
```
We use C++23

avikivity commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+
+* **C++ Standard:** The codebase uses C++20 or newer. Prefer features from these standards, especially coroutines, as they improve code clarity and safety.
+* **Resource Management:** Use Seastar-specific smart pointers and resource management types.
+    * Prefer `foreign_ptr` for sharing resources between Seastar and other code.
```
foreign_ptr is for sharing across shards.

avikivity commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+* **Seastar Framework:** Adhere to the concurrency model and best practices of the Seastar framework. The code should be written to take advantage of Seastar's shard-per-core architecture.
+* **Thread Safety:** Pay close attention to thread safety. Shared data should be protected by appropriate synchronization primitives (e.g., `seastar::mutex`, `std::atomic`), or structured to avoid sharing entirely (e.g., using message passing or per-core data).
+* **`co_await` Semantics:** When using `co_await`, ensure you understand the execution context and potential for context switching.
+* **Avoid Blocking:** Never perform blocking I/O or long-running computations on the reactor thread. Such operations must be offloaded to dedicated thread pools.
```
Break potentially long loops with maybe_yield to avoid reactor stalls.

avikivity commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+* **Zero-Overhead Abstractions:** Prefer abstractions that have zero or minimal runtime overhead.
+* **Avoid Dynamic Allocation:** Minimize dynamic memory allocations, especially in hot code paths. Prefer stack-based or pre-allocated objects.
+* **Compiler Optimizations:** Write code that is easy for the compiler to optimize. Avoid aliasing, use `constexpr` where possible, and prefer standard library algorithms that can be vectorized.
+* **Benchmarking:** Any performance-critical change should be accompanied by benchmarks to validate its impact.
```
Use fragmented containers (utils::chunked_vector) over std::vector where there is potential to break the 128k barrier.

nyh commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+* **Headers:**
+    * Use `#pragma once` for include guards.
+    * Include project headers first, followed by third-party library headers, and finally standard library headers.
+* **No `using namespace std;`:** Never include `using namespace std;` in header files. In source files, use it sparingly and only within function scope to minimize naming conflicts.
```
While at it, you can tell the bot that the "using namespace seastar" is always done and can be assumed. It can usually be assumed because all our header files include seastarx.hh, which contains "using namespace seastar".

avikivity replied:
It's like to move off it, and only import very common names like `future`.

nyh replied:
Yes, I also prefer to explicitly use "seastar" (this is what I recommended in seastar/doc/tutorial.md), but in the Scylla source code we're already hooked on this "using namespace seastar" ;-)

nyh commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+* **Resource Management:** Use Seastar-specific smart pointers and resource management types.
+    * Prefer `foreign_ptr` for sharing resources between Seastar and other code.
+    * Use `lw_shared_ptr` and `shared_ptr` for shared ownership within the Seastar framework.
+    * Avoid `std::unique_ptr`, `std::shared_ptr`, and raw pointers unless there is a clear, documented reason (e.g., interfacing with C-style APIs).
```
A good piece of advice is "Never use the thread-safe but slow std::shared_ptr, use seastar::shared_ptr (aliased to just "shared_ptr") instead."

nyh commented on .copilot/cpp_instructions.md:
````diff
@@ -0,0 +1,88 @@
+      }
+      ```
+* **Integration Tests (Python):** If a native C++ unit test cannot be easily created, offer a Python integration test. This test should operate at a higher level, interacting with the system as a black box (e.g., through a command-line interface or a network protocol).
+    * The Python test should be in the format currently used in the codebase, likely using `pytest` or a custom framework. The test should start the C++ executable, interact with it, and assert the correct behavior.
````
When you write a new test, you do not "start the C++ executable", the test framework already does this for you.
Some test directories include a README.md file which an inquisitive AI (or human!) can read to understand how these tests worked, such as test/cqlpy/README.md

nyh commented on .copilot/general_instructions.md:
```diff
@@ -0,0 +1,119 @@
+
+* **Follow Established Style:** When working on existing files, adhere to the established style of that file. This overrides any personal preference.
+* **Modern C++ and Python:** Favor modern language features that improve clarity and safety.
+    * **C++:** C++17 or C++20 is the standard. Prefer `std::optional`, `std::variant`, and smart pointers (`std::unique_ptr`, `std::shared_ptr`) for resource management.
```
C++23

nyh commented on .copilot/general_instructions.md:
```diff
@@ -0,0 +1,119 @@
+* **Follow Established Style:** When working on existing files, adhere to the established style of that file. This overrides any personal preference.
+* **Modern C++ and Python:** Favor modern language features that improve clarity and safety.
+    * **C++:** C++17 or C++20 is the standard. Prefer `std::optional`, `std::variant`, and smart pointers (`std::unique_ptr`, `std::shared_ptr`) for resource management.
+    * **Python:** Follow PEP 8 and use modern Python features (e.g., f-strings, type hints).
```
Only do this in directories which do this.
test/cqlpy and test/alternator, for example, focus writing a thousand independent small tests, not writing a huge library, so eschew unnecessary features that only make coding more difficult, such as type hints and docstrings (comments look better).

nyh commented on .copilot/general_instructions.md:
```diff
@@ -0,0 +1,119 @@
+# 5. Documentation and Comments
+
+* **Docstrings:** Public APIs (classes, functions) in C++ headers and Python modules should be documented with a brief explanation of their purpose, parameters, and return values.
+* **In-Code Comments:** Comments should be used as a last resort. If a comment is needed to explain a piece of code, it may be a sign that the code is too complex and should be simplified or refactored.
```
I don't agree with this statement, but I know some other people do.
This line also contradicts the next one - the next line says it's good to write a comment to "explain a non-obvious algorithm", but this line says that it's "a sign that the code is too complex and should be simplified or refactored". An algorithm can be non-obvious without needing to be simplified or refactored.

avikivity replied:
Agree with @nyh. Sometimes comments can be abused to explain convoluted code which should otherwise be refactored. But sometimes they explain a complex situation or algorithm.

soyacz commented:
> The more complex instructions were added in the second commit, where I asked Copilot to investigate the whole workspace and try to come up with some additional recommendations based on that research, but also not completely sure if we really need them and if they even couldn't hinder the results (as you say, it might be too much for the Copilot context). The testing that I was doing was with the first commit which was more lean.

This output can be used as reference, feel free to remove things that are redundant

> 
> However on top of that first commit, when trying to work with the Copilot I found out that it would be nice to learn Copilot (i.e., make it part of the instructions) some more specific rules on how to build the codebase and how to run the tests - the Copilot was often able to figure those out even without that, but it required some additional prompting (like e.g. telling it to use a specific test.py command to run the tests, but then it was able to learn it and also adapt for running other tests etc.).
> 
> However, although the information about the build and running tests is there now, I still see the Copilot being confused about how to build the right binaries and run the tests the correct way, so I'll need to do some more testing in that regard to make sure Copilot understands these rules properly (well, they are quite complex indeed).

I'm not sure if this is the issue, but file convention is `<name>.instructions.md` e.g. `.github/instructions/test.instructions.md` - try this, maybe it wasn't taken into account.

another approach would be to create instructions files as you had in .copilot dir and add info to `.github/copilot-instructions.md` to read these (with a hint for every file what it explains and when to use it) when approaching a task .

emdroid commented on .copilot/cpp_instructions.md:
```diff
@@ -0,0 +1,88 @@
+* **Run Relevant Tests:** Following a successful build, run the corresponding unit or integration tests to verify the change's correctness and prevent regressions.
+    * Due to the complexity of the codebase, determining the exact tests to run can be challenging. A good heuristic is to run tests from the same module as the modified code.
+    * **Example commands for running tests:**
+        * `./test.py --no-gather-metric --mode debug cluster/test_topology_ops` (for a specific Python integration test)
```
yeah, copied that one from the command line, but apparently this is a wrong example ... will change it to dev for consistency

avikivity commented on .github/copilot-instructions.md:
````diff
@@ -0,0 +1,101 @@
+```bash
+# Wrap any command with dbuild for old branches lacking dependencies
+./tools/toolchain/dbuild ./configure.py --mode=<mode>
+./tools/toolchain/dbuild ninja <mode>-build
````
And why legacy?

avikivity commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,102 @@
+---
+applyTo: "**/*.{cc,hh,cpp,h}"
```
We don't have .cpp files

emdroid replied:
Removed, kept just .cc and .hh (we have just a single .h file in utils/date.h, but that also seems to be a third party file that isn't expected to be modified very often).

avikivity replied:
Yes, we can probably replace it with the standard calendar library in C++20.

avikivity commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,102 @@
+- Use `seastar::future<T>` for all async operations
+- Prefer coroutines (`co_await`, `co_return`) over `.then()` chains for readability
+- In hot paths where futures are ready, continuations may be more efficient than coroutines
+- Chain futures with `.then()`, never block with `.get()`
```
.get() is okay for seastar::thread

emdroid replied:
updated

avikivity commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,102 @@
+- Prefer coroutines (`co_await`, `co_return`) over `.then()` chains for readability
+- In hot paths where futures are ready, continuations may be more efficient than coroutines
+- Chain futures with `.then()`, never block with `.get()`
+- Use `seastar::do_with()` for temporary state in async chains
```
Anytime a do_with() is there, a coroutine would be better

emdroid replied:
updated

avikivity commented on .github/instructions/cpp.instructions.md:
````diff
@@ -0,0 +1,102 @@
+```
+
+## Error Handling
+- Throw exceptions for errors (futures propagate them automatically)
````
In the data path, avoid throwing exceptions

avikivity replied:
(use boost::outcome/std::expected instead)

emdroid replied:
yes, added a guidance for that

emdroid commented on .github/copilot-instructions.md:
````diff
@@ -0,0 +1,101 @@
+```
+
+**Build Parallelization:**
+- Environment variables `CCACHE_PREFIX=icecc` or `CCACHE_PREFIX=distcc` indicate distributed build setup
````
Right, the intention was to learn copilot how to determine distributed builds. But this can probably be resolved differently (e.g. by a ninja wrapper script, or just prompting the Copilot how many jobs it should use when asking it to run tests). Removed the section.

soyacz commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,105 @@
+- Avoid copies: prefer move semantics
+- Use `utils::chunked_vector` instead of `std::vector` for large allocations (>128KB)
+- Minimize dynamic allocations in hot paths
+- Profile before optimizing
```
I'm not sure if this is relevant instruction for AI

soyacz commented on .github/instructions/python.instructions.md:
```diff
@@ -0,0 +1,74 @@
+## Naming
+- `snake_case` for functions, methods, variables
+- Test files: `test_*.py`
+- Test functions: `test_descriptive_name`
```
better put what should be in the descritpion

soyacz reviewed (commented):
I think this is a good direction.
Better to include less and later improve when you see common mistakes AI does.

emdroid commented on .github/instructions/python.instructions.md:
```diff
@@ -0,0 +1,74 @@
+
+## Naming
+- `snake_case` for functions, methods, variables
+- Test files: `test_*.py`
```
removed

emdroid commented:
New version:
* added recommendations for simplicity bias, and not blindly implement requests (to consider different approaches)
* updated the C++ formatting section (removed the .clang-format reference to not have to read extra files, added some rules from the Seastar coding style)
* removed unnecessary coding style guidance for Python

[diff](https://github.com/scylladb/scylladb/compare/eb0aa41cd951e82df3a28a0c20682254ba38cac3..f146a51154ccc4653843129b7f8bd0d29e043086)

nyh commented on .github/copilot-instructions.md:
````diff
@@ -0,0 +1,92 @@
+
+### Python Integration Tests
+```bash
+# Only requires Scylla binary (full build NOT needed)
````
Sadly, some specific tests do require compiling additional things like wasm stuff. But if you run only specific tests and not the entire suite, indeed just compiling "scylla" is usually enough.

nyh commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,92 @@
+- Consider different approaches, weigh pros and cons, and recommend the best fit for the specific context
+
+## Logging
+Use structured logging in all languages that support it. Follow these guidelines:
```
What "languages" are we talking about?
Scylla is C++.
Tests don't need any sophisticated "logging". They can print output in whatever way they wish, and if the test fails - you get that output. It's as simple as that.

nyh commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,107 @@
+- Use `auto` when type is obvious from RHS
+- Avoid `auto` when it obscures the type
+- Use range-based for loops: `for (const auto& item : container)`
+- Prefer algorithms over raw loops when clear
```
Personally, I would use the opposite rule: Prefer loops over "algorithms" when clear.
If you can replace a 10 line loop by a single algorithm, great. But if you can replace a clear 3-line loop (like an obvious loop over elements of an array) by a piping of 3 algorithms - no, don't do this.

nyh commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,107 @@
+- Avoid raw pointers except for non-owning references or C API interop
+
+## Seastar Asynchronous Programming
+- Use `seastar::future<T>` for all async operations
```
I see you say that later.

nyh commented on .github/instructions/python.instructions.md:
```diff
@@ -0,0 +1,66 @@
+Never use `from module import *`
+
+## Test Structure
+- Use descriptive names that convey intent: `test_<action>_<expected_result>`
```
I think this instruction is arbitrary, and doesn't match real-world needs. Just as a random example from the first file I looked at, consider test/alternator/query.py (the entire test file is dedicated to the Query operation), test_query_sort_order_string - it tests the sort order returned by a query, when the sort key is the string. There is no real "action" or "expected result" in this test's name.

It is important to emphasize that although the name "test_query_sort_order_string" should be suggestive, and unique (so you can "git grep" a test name), there is no reason for it to be somehow self-explanatory. It is almost never possible to guess exactly what a test does just from its name. That's why a comment in front of every test is very important. For this example test we have:
```python
# Items returned by Query should be sorted by the sort key. The following
# tests verify that this is indeed the case, for the three allowed key types:
# strings, binary, and numbers. These tests test not just the Query operation,
# but inherently that the sort-key sorting works.
def test_query_sort_order_string(test_table):
```

Note also how this is an example of the point I made in a comment below - comments are better than docstrings when it comes to tests, because they also allow explaining something about a whole set of functions, not just one.

emdroid replied:
Makes sense, removed the template (kept just "use descriptive names that convey intent", which should be enough I think, plus that docstring or comment description should be used).

emdroid commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,92 @@
+- To find which binary contains a test, check `configure.py` in the repository root (primary source) or `test/<suite>/CMakeLists.txt`
+- To rebuild a specific test binary: `ninja -j150 build/<mode>/test/<suite>/<binary_name>`
+- Examples: 
+  - `ninja -j150 build/dev/test/boost/combined_tests` (contains group0_voter_calculator_test.cc and others)
```
Good catch, removed. This was a remainder from the previous version that included distributed builds (where "-j150" can be used), but in the usual case ninja selects the best parallelism indeed, so it's better to not use explicit job settings. The case of distributed build could be handled separately, e.g. by a ninja wrapper script (outside of scylla), or in the specific Copilot prompt.

emdroid commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,107 @@
+- `printf` family (use logging or fmt)
+- Raw pointers for ownership
+- `using namespace` in headers
+- Blocking operations (`sleep`, `read`, `std::mutex`, `std::atomic`)
```
Yes, updated

emdroid commented on .github/instructions/cpp.instructions.md:
```diff
@@ -0,0 +1,107 @@
+## Testing
+When modifying existing code, follow TDD: create/update test first, then implement.
+- Examine existing tests for style and structure
+- Use Boost.Test framework
```
As this is for C++ instructions, I suppose we only use Boost.test for the unit tests there

emdroid commented on .github/instructions/python.instructions.md:
````diff
@@ -0,0 +1,66 @@
+
+```python
+def test_select_returns_inserted_data(cql, test_keyspace):
+    """Brief description of what is tested."""
````
Yes, added to the testing section (as I removed this one).

I suppose the general docstring preference is because it is more canonical for Python, and also then it is handled in the IDEs (like VS Code) ... although it maybe will show comments as well (not sure right now). but changed the wording to include either docstring or comments (and there is also the guidance to follow the current file/dir conventions).

emdroid commented on .github/instructions/python.instructions.md:
```diff
@@ -0,0 +1,66 @@
+def test_select_returns_inserted_data(cql, test_keyspace):
+    """Brief description of what is tested."""
+    # Arrange: setup schema and data
+    cql.execute(f"CREATE TABLE {test_keyspace}.tbl (id int PRIMARY KEY, data text)")
```
I was actually thinking about removing the example even before, so removed now (there are enough examples in the codebase for the Copilot anyway).
